cache/ringbuffer.go

```diff
@@ -21,46 +21,74 @@ import (
 )
 
 type ringBuffer struct {
-	buffer []*clientv3.Event
+	buffer []*batch
 	// head is the index immediately after the last non-empty entry in the buffer (i.e., the next write position).
 	head, tail, size int
 }
 
+// batch groups all events that share one ModRevision.
+type batch struct {
+	rev    int64
+	events []*clientv3.Event
+}
+
 type KeyPredicate = func([]byte) bool
 
 func newRingBuffer(capacity int) *ringBuffer {
 	// assume capacity > 0 – validated by Cache
 	return &ringBuffer{
-		buffer: make([]*clientv3.Event, capacity),
+		buffer: make([]*batch, capacity),
 	}
 }
 
-func (r *ringBuffer) Append(event *clientv3.Event) {
-	if r.size == len(r.buffer) { // full → overwrite oldest
+func (r *ringBuffer) Append(eventBatch []*clientv3.Event) {
+	if len(eventBatch) == 0 {
+		panic("ringBuffer.Append: empty group")
+	}
+	rev := eventBatch[0].Kv.ModRevision
+	for i, ev := range eventBatch[1:] {
+		if ev.Kv.ModRevision != rev {
+			panic(fmt.Sprintf("ringBuffer.Append: mixed revisions at idx %d (%d vs %d)", i+1, ev.Kv.ModRevision, rev))
+		}
+	}
+	b := &batch{rev: rev, events: eventBatch}
+
+	if r.size == len(r.buffer) {
 		r.tail = (r.tail + 1) % len(r.buffer)
 	} else {
 		r.size++
 	}
-
-	r.buffer[r.head] = event
+	r.buffer[r.head] = b
 	r.head = (r.head + 1) % len(r.buffer)
 }
 
+func (r *ringBuffer) AppendMixed(events []*clientv3.Event) {
+	for i := 0; i < len(events); {
+		rev := events[i].Kv.ModRevision
+		j := i + 1
+		for j < len(events) && events[j].Kv.ModRevision == rev {
+			j++
+		}
+		r.Append(events[i:j])
+		i = j
+	}
```

Comment:
Previous implementation was clearer. Don't use nested loops if not needed. They imply O(n^2) complexity if you make mistake.

If you want to avoid allocations just use the following code:
```
	start := 0
	for end := 1; end < len(events); end++ {
		if events[end].Kv.ModRevision != events[start].Kv.ModRevision {
			r.Append(events[start:end])
			start = end
		}
	}
	if start < len(events) {
		r.Append(events[start:])
	}
```